data/dfinsupp.lean

```diff
@@ -0,0 +1,868 @@
+import data.finset data.set.finite algebra.big_operators algebra.module algebra.pi_instances
+
+universes u u₁ u₂ v v₁ v₂ v₃ w x y l
+
+variables (ι : Type u) (β : ι → Type v)
+
+instance decidable_zero_symm {γ : Type w} [has_zero γ] [decidable_pred (eq (0 : γ))] : decidable_pred (λ x, x = (0:γ)) :=
+λ x, decidable_of_iff (0 = x) eq_comm
+
+namespace dfinsupp
+
+@[reducible] protected def finset (s : finset ι) : Type* :=
+Π i : (↑s : set ι), β i.1

```

Comment:
We don't need partial functions here. They can be totalized with the value zero outside the finset. Then the quotient is just over different support sets.

Reply:
here we don't have decidable equality on the indexing set, yet.

Reply:
I know, but it doesn't matter. You can just define the set of functions zero outside a range, and quotient over what that range is. That way you get `app` without decidable equality, but you need decidable equality to extract `support`, which is the way it should be.

Reply:
then how do you even define "single"?

Reply:
You can't define single without decidable equality, but I don't know any representation that would allow this. If you just say "1 on {x}" and say it is partial so you don't have to define the rest, then it's not really a single function since you can't combine these into really any combinations. There is certainly no "double" function definable without decidable equality.

Reply:
If the domain of the functions is something like computable `real`, then of course there are no definable finitely supported functions anyway, but the collection of all of them still has nice closure properties without decidable equality.

Reply:
what are those closure properties you mention?

Reply:
Closure under addition, scalar multiplication, generally applying finitary functions with appropriate conditions.

Another representation of `A ->0 B` that is probably more constructive friendly is a multiset of (A, B) pairs, where we view any two values with the same A coordinate as being implicitly summed up. That is, there is a quotient that identifies (a, b1) :: (a, b2) :: s with (a, b1 + b2) :: s. This way you can calculate with the values even if the underlying set doesn't have decidable equality, because it's not actually a function from A to B.

Reply:
In [nlab](https://ncatlab.org/nlab/show/direct+sum) at least, the coproduct is distinguished from the direct sum, in that elements of coproduct are not functions, i.e. the coproduct is not identified with a subset of the product (the product = Π i, β i).

Also, I don't think you can define addition, because unfortunately with this implementation of finset, union requires decidable equality.

Reply:
How does this look?
```lean
structure pre : Type (max u v) :=
(to_fun : Π i, β i)
(pre_support : multiset ι)
(zero : ∀ i ∉ pre_support, to_fun i = 0)

instance : setoid (pre ι β) :=
{ r := λ x y, ∀ i, x.to_fun i = y.to_fun i,
  iseqv := ⟨λ f i, rfl, λ f g H i, (H i).symm,
    λ f g h H1 H2 i, (H1 i).trans (H2 i)⟩ }
```

Reply:
Addition is just addition of multisets with that representation. You aren't trying to keep it minimal so there aren't any decidable equality things that come up.

Reply:
maybe on R with fixed endpoints, but can you define a function such that `f(x)=0` and `f(1)=1`, where `x` is fixed and arbitrary not equal to 1? Or I can up the difficulty by taking an equivalence relation on R identifying all the rationals and all the irrationals, and define `f(rational) = 0` and `f(irrational) = 1`.

Reply:
but I'm talking about using partial functions to define dfinsupp

Reply:
Yes, I know. I'm proposing you take R viewed as a quotient that identifies all rationals with each other and all irrationals (so it is bool classically), and construct the finitely supported function on this set that takes the value `1` at the rational point and `2` at the irrational point. (I shifted the values away from zero so you can't leave it undefined at either one.)